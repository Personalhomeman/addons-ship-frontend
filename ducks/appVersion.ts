import { Dispatch } from 'redux';
import { createAction, createReducer } from 'deox';

import api from '@/services/ship-api';
import { actionTypeCreator } from '@/utils';
import { uploadFileToS3 } from '@/utils/file';
import { AppVersion } from '@/models';
import { Uploadable } from '@/models/uploadable';
import { RootState } from '@/store';

const $ = actionTypeCreator('APP_VERSION');

const _fetchAppVersion = (appSlug: string, versionId: string) => async (
  dispatch: Dispatch,
  getState: () => RootState
) => {
  const {
    auth: { token }
  } = getState();

  api.setToken(token);
  dispatch(fetchAppVersion.next());

  try {
    const appVersion = await api.getAppVersion(appSlug, versionId);

    dispatch(fetchAppVersion.complete(appVersion));
  } catch (error) {
    dispatch(fetchAppVersion.error(error));
  }
};

export const fetchAppVersion = Object.assign(_fetchAppVersion, {
  next: createAction($`GET_NEXT`),
  complete: createAction($`GET_COMPLETE`, resolve => (appVersion: AppVersion) => resolve(appVersion)),
  error: createAction($`GET_ERROR`, resolve => (error: Error) => resolve(error))
});

const _updateAppVersion = (appVersion: AppVersion) => async (dispatch: Dispatch, getState: () => RootState) => {
  const {
    auth: { token }
  } = getState();

  api.setToken(token);

  dispatch(updateAppVersion.next());

  try {
    const newAppVersion = await api.updateAppVersion(appVersion);

    dispatch(updateAppVersion.complete(newAppVersion));
  } catch (error) {
    dispatch(updateAppVersion.error(error));
  }
};

export const updateAppVersion = Object.assign(_updateAppVersion, {
  next: createAction($`UPDATE_NEXT`),
  complete: createAction($`UPDATE_COMPLETE`, resolve => (appVersion: AppVersion) => resolve(appVersion)),
  error: createAction($`UPDATE_ERROR`, resolve => (error: Error) => resolve(error))
});

// Upload screenshots
const _uploadScreenshots = (appSlug: string, versionId: string, screenshots: Uploadable[], files: File[]) => async (
  dispatch: Dispatch,
  getState: () => RootState
) => {
  const {
    auth: { token }
  } = getState();

  api.setToken(token);

  dispatch(uploadScreenshots.next());

  try {
    // Get presigned Urls
    const withPresignedUrl = await api.uploadScreenshots(appSlug, versionId, screenshots);

    // Upload images to S3
    await Promise.all(
      withPresignedUrl.map(({ filename, filesize, uploadUrl }) => {
        const file = files.find(({ name, size }) => name === filename && size === filesize);

        return uploadFileToS3(file, uploadUrl);
      })
    );

    // Mark screenshots as uploaded
    await api.uploadedScreenshots(appSlug, versionId);
    dispatch(uploadScreenshots.complete());
  } catch (error) {
    dispatch(uploadScreenshots.error(error));
  }
};

export const uploadScreenshots = Object.assign(_uploadScreenshots, {
  next: createAction($`UPLOAD_SCREENSHOTS_NEXT`),
  complete: createAction($`UPLOAD_SCREENSHOTS_COMPLETE`),
  error: createAction($`UPLOAD_SCREENSHOTS_ERROR`, resolve => (error: Error) => resolve(error))
});

const _publishAppVersion = (appVersion: AppVersion) => async (dispatch: Dispatch, getState: () => RootState) => {
  const {
    auth: { token }
  } = getState();

  api.setToken(token);

  dispatch(publishAppVersion.next());

  try {
    await api.publishAppVersion(appVersion);

    dispatch(publishAppVersion.complete());
  } catch (error) {
    dispatch(publishAppVersion.error(error));
  }
};

export const publishAppVersion = Object.assign(_publishAppVersion, {
  next: createAction($`PUBLISH_NEXT`),
  complete: createAction($`PUBLISH_COMPLETE`, resolve => () => resolve()),
  error: createAction($`PUBLISH_ERROR`, resolve => (error: Error) => resolve(error))
});

const defaultState: AppVersionState = { appVersion: null, isPublishInProgress: false };

export type AppVersionState = {
  appVersion: AppVersion | null;
  isPublishInProgress?: boolean;
};
export const appVersion = createReducer(defaultState, handleAction => [
  handleAction([fetchAppVersion.complete, updateAppVersion.complete], (state, { payload }) => ({
    ...state,
    appVersion: payload
  })),
  handleAction(publishAppVersion.next, state => ({ ...state, isPublishInProgress: true })),
  handleAction([publishAppVersion.complete, publishAppVersion.error], state => ({
    ...state,
    isPublishInProgress: false
  }))
]);
