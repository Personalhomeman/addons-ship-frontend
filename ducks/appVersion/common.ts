import { actionTypeCreator } from '@/utils';
export const $ = actionTypeCreator('APP_VERSION');
