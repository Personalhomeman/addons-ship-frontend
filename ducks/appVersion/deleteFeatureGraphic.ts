import { Dispatch } from 'redux';
import { createAction } from 'deox';

import { ShipAPIService } from '@/services/ship-api';
import { RootState } from '@/store';

import { $ } from './common';

const _deleteFeatureGraphic = (appSlug: string, versionId: string) => async (
  dispatch: Dispatch,
  _getState: () => RootState,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(deleteFeatureGraphic.next());

  try {
    await shipApi.deleteFeatureGraphic(appSlug, versionId);

    dispatch(deleteFeatureGraphic.complete());
  } catch (error) {
    console.error('deleteFeatureGraphic', error);
    dispatch(deleteFeatureGraphic.error(error));
  }
};

const deleteFeatureGraphic = Object.assign(_deleteFeatureGraphic, {
  next: createAction($`DELETE_FEATURE_GRAPHIC_NEXT`),
  complete: createAction($`DELETE_FEATURE_GRAPHIC_COMPLETE`),
  error: createAction($`DELETE_FEATURE_GRAPHIC_ERROR`, resolve => (error: Error) => resolve(error))
});

export default deleteFeatureGraphic;
