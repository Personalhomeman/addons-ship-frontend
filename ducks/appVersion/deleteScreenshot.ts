import { Dispatch } from 'redux';
import { createAction } from 'deox';

import { ShipAPIService } from '@/services/ship-api';
import { RootState } from '@/store';

import { $ } from './common';

const _deleteScreenshot = (appSlug: string, versionId: string, screenshotId: string) => async (
  dispatch: Dispatch,
  _getState: () => RootState,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(deleteScreenshot.next());

  try {
    await shipApi.deleteScreenshot(appSlug, versionId, screenshotId);

    dispatch(deleteScreenshot.complete());
  } catch (error) {
    console.error(deleteScreenshot, error);
    dispatch(deleteScreenshot.error(error));
  }
};

const deleteScreenshot = Object.assign(_deleteScreenshot, {
  next: createAction($`DELETE_SCREENSHOTS_NEXT`),
  complete: createAction($`DELETE_SCREENSHOTS_COMPLETE`),
  error: createAction($`DELETE_SCREENSHOTS_ERROR`, resolve => (error: Error) => resolve(error))
});

export default deleteScreenshot;
