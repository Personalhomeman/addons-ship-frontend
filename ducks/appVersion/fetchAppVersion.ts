import { Dispatch } from 'redux';
import { createAction } from 'deox';

import { ShipAPIService } from '@/services/ship-api';
import { AppVersion } from '@/models';
import { RootState } from '@/store';

import { $ } from './common';
import { RequestError } from '@/models/errors';

const _fetchAppVersion = (appSlug: string, versionId: string) => async (
  dispatch: Dispatch,
  _getState: () => RootState,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(fetchAppVersion.next());

  try {
    const appVersion = await shipApi.getAppVersion(appSlug, versionId);
    appVersion.screenshotDatas = await shipApi.getScreenshots(appSlug, versionId);

    try {
      appVersion.featureGraphicData = await shipApi.getFeatureGraphic(appSlug, versionId);
    } catch (error) {
      if (!(error instanceof RequestError && error.status === 404)) {
        throw error;
      }
    }

    dispatch(fetchAppVersion.complete(appVersion));
  } catch (error) {
    console.error('fetchAppVersion', error);
    dispatch(fetchAppVersion.error(error));
  }
};

const fetchAppVersion = Object.assign(_fetchAppVersion, {
  next: createAction($`GET_NEXT`),
  complete: createAction($`GET_COMPLETE`, resolve => (appVersion: AppVersion) => resolve(appVersion)),
  error: createAction($`GET_ERROR`, resolve => (error: Error) => resolve(error))
});

export default fetchAppVersion;
