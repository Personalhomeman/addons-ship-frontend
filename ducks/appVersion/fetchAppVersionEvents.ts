import { Dispatch } from 'redux';
import { createAction } from 'deox';

import api from '@/services/ship-api';
import { AppVersionEvent } from '@/models';

import { $ } from './common';

const _fetchAppVersionEvents = (appSlug: string, versionId: string) => async (dispatch: Dispatch) => {
  dispatch(fetchAppVersionEvents.next());

  try {
    const appVersionEvents = await api.getAppVersionEvents(appSlug, versionId);

    dispatch(fetchAppVersionEvents.complete(appVersionEvents));
  } catch (error) {
    dispatch(fetchAppVersionEvents.error(error));
  }
};

const fetchAppVersionEvents = Object.assign(_fetchAppVersionEvents, {
  next: createAction($`GET_EVENTS_NEXT`),
  complete: createAction($`GET_EVENTS_COMPLETE`, resolve => (appVersionEvents: AppVersionEvent[]) => resolve(appVersionEvents)),
  error: createAction($`GET_EVENTS_ERROR`, resolve => (error: Error) => resolve(error))
});

export default fetchAppVersionEvents;
