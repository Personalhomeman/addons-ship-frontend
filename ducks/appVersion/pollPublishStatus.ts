import { createAction, ActionType, ofType } from 'deox';
import { Observable, timer } from 'rxjs';
import { exhaustMap, map, takeUntil, mergeMap } from 'rxjs/operators';
import getConfig from 'next/config';

import { RootState } from '@/store';
import { ShipAPIService } from '@/services/ship-api';
import { AppVersion, AppVersionEvent } from '@/models';

import { $ } from './common';

const {
  publicRuntimeConfig: { POLL_INTERVAL }
} = getConfig();
const pollInterval = Number(POLL_INTERVAL) || 5000;

const pollPublishStatus = {
  start: createAction($`POLL_START`, resolve => (appVersion: AppVersion) => resolve(appVersion)),
  cancel: createAction($`POLL_CANCEL`),
  error: createAction($`POLL_ERROR`, resolve => (error: Error) => resolve(error)),
  complete: createAction($`POLL_COMPLETE`, resolve => (events: AppVersionEvent[]) => resolve(events))
};

export type PollPublishStatusAction = ActionType<
  | typeof pollPublishStatus.start
  | typeof pollPublishStatus.cancel
  | typeof pollPublishStatus.error
  | typeof pollPublishStatus.complete
>;

export const pollPublishStatusEpic = (
  action$: Observable<PollPublishStatusAction>,
  _state$: Observable<RootState>,
  { shipApi }: { shipApi: ShipAPIService }
) =>
  action$.pipe(
    ofType(pollPublishStatus.start),
    exhaustMap(({ payload }) =>
      timer(0, pollInterval).pipe(
        // Should try this with `backoff-rxjs/intervalBackoff`
        mergeMap(() => shipApi.getAppVersionEvents(payload.appSlug, payload.id as string)),
        map(events => pollPublishStatus.complete(events)),
        takeUntil(action$.pipe(ofType(pollPublishStatus.cancel)))
      )
    )
  );

export default pollPublishStatus;
