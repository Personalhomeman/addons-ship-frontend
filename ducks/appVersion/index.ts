import { createReducer } from 'deox';

import { AppVersion, AppVersionEvent } from '@/models';

import fetchAppVersion from './fetchAppVersion';
import updateAppVersion from './updateAppVersion';
import uploadScreenshots from './uploadScreenshots';
import deleteScreenshot from './deleteScreenshot';
import uploadFeatureGraphic from './uploadFeatureGraphic';
import publishAppVersion from './publishAppVersion';
import pollPublishStatus, { pollPublishStatusEpic } from './pollPublishStatus';
import fetchAppVersionEvents from './fetchAppVersionEvents';

export type AppVersionState = {
  appVersion: AppVersion | null;
  isPublishInProgress?: boolean;
  events: AppVersionEvent[];
};

const defaultState: AppVersionState = { appVersion: null, isPublishInProgress: false, events: [] };

export {
  fetchAppVersion,
  updateAppVersion,
  uploadScreenshots,
  deleteScreenshot,
  uploadFeatureGraphic,
  publishAppVersion,
  pollPublishStatus,
  pollPublishStatusEpic
};
export default createReducer(defaultState, handleAction => [
  handleAction([fetchAppVersion.complete, updateAppVersion.complete], ({ appVersion, ...state }, { payload }) => ({
    ...state,
    appVersion: { ...appVersion, ...payload }
  })),
  handleAction(publishAppVersion.next, state => ({ ...state, isPublishInProgress: true })),
  handleAction([publishAppVersion.complete, publishAppVersion.error], state => ({
    ...state,
    isPublishInProgress: false
  })),
  handleAction([pollPublishStatus.complete, fetchAppVersionEvents.complete], (state, { payload }) => ({
    ...state,
    events: payload
  }))
]);
