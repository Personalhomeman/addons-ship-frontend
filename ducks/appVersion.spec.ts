jest.mock('@/services/ship-api');

import configureMockStore, { MockStoreCreator } from 'redux-mock-store';
import thunk from 'redux-thunk';

import { fetchAppVersion } from './appVersion';

describe('appVersion', () => {
  let mockStore: MockStoreCreator;
  beforeEach(() => {
    mockStore = configureMockStore([thunk]);
  });

  it('should update the state', async () => {
    const store = mockStore();

    await store.dispatch(fetchAppVersion('app-slug', 'version-id') as any);

    expect(store.getActions()).toMatchSnapshot();
  });
});
