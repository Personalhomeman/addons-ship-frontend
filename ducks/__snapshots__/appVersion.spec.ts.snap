// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appVersion should update the state 1`] = `
Array [
  Object {
    "type": "APP_VERSION_GET_NEXT",
  },
  Object {
    "type": "APP_VERSION_GET_COMPLETE",
  },
]
`;
